templates/README.md

```diff
@@ -0,0 +1,221 @@
+    # and let the user filter down the available templates to the one
+    # the user actually wants:
+    # A single selected template only means no conflicts and no smart
+    # merging code. This has to be done using labels to allow efficient
+    # searching.
+    # The format has the following meaning:
+    # {os,flavor,size}.template.cnv.io/{value}: {version} or just true
+    os.template.cnv.io/windows2k12r2: 201809 
+    os.template.cnv.io/windows8: 201809 
+    os.template.cnv.io/windows7: 201809 
+    flavor.template.cnv.io/minimal: 201809
+    # io.cnv.vm.size not specified means all?
+
+# Parameters must come from a subset of well known names
+# so the UI can properly work with those.
+parameters:
+- description: VM name
+  from: '[A-Za-z0-9]{1,16}'
+  generate: expression
+  name: VM_NAME
+- description: Memory size
+  from: '[A-Z0-9]{8}'
+  generate: expression
+  name: MEMORY_SIZE
+  value: 1Gi
+
+objects:
+# The full VM template with placeholders for either scalars like memory
+# or yaml structures like disks
+- apiVersion: v1
+  kind: VirtualMachine
+  metadata:
+    name: ${VM_NAME}
+  spec:
+    domain:
+      resources:
+        initial:
+          memory: ${MEMORY_SIZE}Mi
+      devices:
+        disks:
+        # This should be interpreted as a template disk by the UI,
+        # thanks to the template.cnv.io/default annotations
+        # This must still result in a bootable VM when used as is.
+        # This way we can both use Templates for creating new VM as well
+        # as for converting an existing VM to a template
+        - name: default-disk
+          disk:
+            dev: vda
+          volumeName: default-volume
+
+        interfaces:
+        # This should be interpreted as a template network by the UI,
+        # thanks to the template.cnv.io/default annotations
+        # This must still result in a bootable VM when used as is.
+        # This way we can both use Templates for creating new VM as well
+        # as for converting an existing VM to a template
+        - type: pod-network
+          name: default-network
+          model:
+          type: virtio
+
+    volumes:
+        # This should be interpreted as a template volume by the UI,
+        # thanks to the template.cnv.io/default annotations
+        # This must still result in a bootable VM when used as is.
+        # This way we can both use Templates for creating new VM as well
+        # as for converting an existing VM to a template
+      - name: default-volume
+        registryDisk:
+          image: test/image
+
+```
+
+### VM.yaml
+
+The above template should result in the following VM object:
+
+```YAML
+# vim: sts=2 sw=2 et
+apiVersion: v1
+kind: VirtualMachine
+metadata:
+  name: windows-10-1
+  annotations:
+    # Data used for the UI, ignored by kubevirt
+    # Arbitrary format as needed to be able to
+    # repopulate the UI or the template processor
+    # and get the same output
+    template.cnv.io/original:
```

Comment:
I _think_ I have seen that OCP Templates already store the original parameter values. Somethig to be checked.

Reply:
I will check that, but yes, we want to reuse as much as possible.

Reply:
Actually - It looks like this is done by the UI, or a broker.

But plain temapltes `oc process  | oc apply -f -` will not carry ownership informations.
To be honest, it's not ownership, but object reference would be helpful.

Eventually something to enhance in template component.